src/Storages/MergeTree/MergeTreeSink.cpp

```diff
@@ -44,6 +45,8 @@ MergeTreeSink::~MergeTreeSink()
     if (!delayed_chunk)
         return;
 
+    chassert(isCancelled() || std::uncaught_exceptions());
```

Comment:
This assert was triggered in the CI: https://s3.amazonaws.com/clickhouse-test-reports/72741/10775ef3ecf538dd418e026bcbf7e643098321c9/stateless_tests__tsan__s3_storage__[2_3].html